### PR TITLE
[AIRFLOW-5945] Make inbuilt OperatorLinks work when using Serialization

### PR DESCRIPTION
**Pull request by kaxil**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-5945

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:
The inbuilt Operator links were not working when using DAG Serialization (which was 1 of the limitations of DAG Serialization). The work-around previously was adding those OperatorLinks using Airflow Plugin.

This PR fixes it so we inbuilt links work out of the box. 

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
Added

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### CODE REVIEW

**Comment by kaxil** [comment]
This PR is dependent on #6714 so please only review https://github.com/apache/airflow/pull/6715/commits/92078dab7c201c1fdcd39c6b4850973839fd8d54 for this PR.

**Comment by mik-laj** [comment]
What do you think to save the generated links in a separate column in the database after Task execution?  I think it will be nicer because it will not be storing additional partial data, but the final data that interests us.  This could be realized independent of DAG serialization. In addition, it will simplify the logic of handling these links. Now it is not possible to download all links for a given task, but we download one link in one HTTP request. https://github.com/apache/airflow/blob/master/airflow/www/templates/airflow/dag.html#L492-L518

**Comment by kaxil** [comment]
> What do you think to save the generated links in a separate column in the database after Task execution? I think it will be nicer because it will not be storing additional partial data, but the final data that interests us. This could be realized independent of DAG serialization. In addition, it will simplify the logic of handling these links. Now it is not possible to download all links for a given task, but we download one link in one HTTP request. https://github.com/apache/airflow/blob/master/airflow/www/templates/airflow/dag.html#L492-L518

This is actually the 1st solution I mentioned to Ash and I thought it was the way to go :) Unfortunately Ash mentioned a use-case that ruled it out which was I think "S3 signed URL can have an expiration date of 15 mins and change after that". @ashb can probably explain that in more detail but due to that use-case storing the links in DB wouldn't work.

**Comment by ashb** [comment]
Yeah, we did think about that as a solution (which means we should probably add a comment/"architecture decision record" somewhere) but the issue is that I know of a link I want to add to EMR that isn't fully static: Linking from EmrAddStepOperator to the job logs in S3, which involves signing a URL that will only be valid for 15minutes. (This uses the Airflow servers AWS credentials, rather than needing the user to have the right permissions, which doesn't really work with S3 easily.)

Because of this only allowing static links in the DB would not work for this use case

File: airflow/contrib/operators/qubole_operator.py [hunk 1]
```diff
@@ -240,3 +241,9 @@ def __setattr__(self, name, value):
             self.kwargs[name] = value
         else:
             object.__setattr__(self, name, value)
+
+    @classmethod
+    def get_serialized_fields(cls):
+        """Serialized QuboleOperator contain exactly these fields."""
+        cls._serialized_fields = frozenset(super().get_serialized_fields() | {"qubole_conn_id"})
```
Comment (ashb) [inline airflow/contrib/operators/qubole_operator.py]: To pull across the comments from the closed pr: this should have the `if not cls._serialized_fields` guard here, otherwise we re-calculate this each time it's called (not the end of the world, but still)

Comment (kaxil) [reply airflow/contrib/operators/qubole_operator.py]: The problem is when calling `super().get_serialized_fields()` it would set `cls._serialized_fields` from the BaseOperator, hence it would not have `{"sql"}` in it

File: airflow/serialization/base_serialization.py [hunk 2]
```diff
@@ -216,7 +216,11 @@ def _deserialize(cls, encoded_var):  # pylint: disable=too-many-return-statement
             return [cls._deserialize(v) for v in encoded_var]
 
         assert isinstance(encoded_var, dict)
-        var = encoded_var[Encoding.VAR]
+        try:
+            var = encoded_var[Encoding.VAR]
+        except KeyError as e:
+            logging.info("encoded_var: %s", encoded_var)
```
Comment (ashb) [inline airflow/serialization/base_serialization.py]: Is this just debugging?

Comment (kaxil) [reply airflow/serialization/base_serialization.py]: Yes but I kept it intentionally as it would be useful to log what key was not found.

Happy to remove it if you think

File: airflow/contrib/operators/qubole_operator.py (hunk 1 above)
Comment (ashb) [reply airflow/contrib/operators/qubole_operator.py]: Oh yeah.

There's no point storing it on the class variable then is there? Just `return super().get_serialized_fields() | frozenset(["quoble_conn_id"]}`?

File: airflow/serialization/serialized_baseoperator.py [hunk 3]
```diff
@@ -89,7 +96,10 @@ def deserialize_operator(cls, encoded_op: dict) -> BaseOperator:
                         operator.__module__ == encoded_op["_task_module"]:
                     op_extra_links_from_plugin.update({ope.name: ope})
 
-        setattr(op, "operator_extra_links", list(op_extra_links_from_plugin.values()))
+        # If OperatorLinks is defined in Plugins but not in the Operator
+        # set the Operator links attribute
+        if op_extra_links_from_plugin and "_operator_links_sources" not in encoded_op:
+            setattr(op, "operator_extra_links", list(op_extra_links_from_plugin.values()))
```
Comment (ashb) [inline airflow/serialization/serialized_baseoperator.py]: What if we define Links in a plugin _and_ in the operator?

I'm finding this code (including the existing code) a bit hard to follow. Hmmmm.

File: airflow/contrib/operators/qubole_operator.py (hunk 1 above)
Comment (kaxil) [reply airflow/contrib/operators/qubole_operator.py]: The only reason I stored it is for consistency as otherwise `cls._serialized_fields` will have a different value than `cls.get_serialized_fields()`. We don't use `cls._serialized_fields` anywhere, maybe we can just remove it from the Base Class too than WDYT? If not, let's keep it at both places as the results won't match

File: airflow/serialization/serialized_baseoperator.py (hunk 3 above)
Comment (kaxil) [reply airflow/serialization/serialized_baseoperator.py]: If we define in a plugin and in operator, then the following code get's executed:
https://github.com/apache/airflow/pull/6715/files#diff-f58748b66b2d3d00c8132103faea223fR114-R121

I can add more detailed comments if it is not clear.

File: airflow/serialization/serialized_baseoperator.py [hunk 4]
```diff
@@ -124,3 +142,50 @@ def _is_excluded(cls, var, attrname, op):
...
+        :return: De-Serialized Operator Link
+        """
+
+        # Extra Operator Links
+        op_predefined_extra_links = {}
+
+        for _operator_links_source in encoded_op_links:
+            _operator_link_class, data = list(_operator_links_source.items())[0]
+            try:
+                single_link = import_string(_operator_link_class)
```
Comment (ashb) [inline airflow/serialization/serialized_baseoperator.py]: I'm not sure if we should do this here (the import string) - my first thought was that we should look through the loaded plugins only.

File: airflow/serialization/serialized_baseoperator.py [hunk 5]
```diff
@@ -124,3 +142,50 @@ def _is_excluded(cls, var, attrname, op):
...
+        """
+        Serialize Operator Links. Store the import path of the OperatorLink and the arguments
+        passed to it. Example ``[{'airflow.gcp.operators.bigquery.BigQueryConsoleLink': {}}]``
+
+        :param operator_extra_links: Operator Link
+        :return: Serialized Operator Link
+        """
+        return [
+            {
+                "{}.{}".format(x.__class__.__module__, x.__class__.__name__): cattr.unstructure(x)
```
Comment (ashb) [inline airflow/serialization/serialized_baseoperator.py]: Do we need to document somewhere that Links should use `cattr`/atrrs to be deserialized corectly?

File: airflow/serialization/serialized_baseoperator.py (hunk 4 above)
Comment (kaxil) [reply airflow/serialization/serialized_baseoperator.py]: It is needed for the inbuilt OperatorLinks (BigQuery and Qubole) to work

File: airflow/serialization/serialized_baseoperator.py (hunk 5 above)
Comment (kaxil) [reply airflow/serialization/serialized_baseoperator.py]: That is only needed for inbuilt operators, all the other OperatorLinks can directly be registered via Plugins.

Also, `cattr/atrrs` is only needed when the OperatorLinks are like `BigQueryConsoleIndexableLink` i.e. they have required parameters like `index`. E.g there are not required by `BigQueryConsoleLink` or `QDSLink`

**Review by ashb — COMMENTED** [review]
Somewhere I'd like to see us do a "static" deserialization that includes a Links - both ones defined in the operator, and one registered via a plugin, like we have with `serialized_simple_dag_ground_truth`.

(it might be time to move those to JSON files rather than having them in the test file? Don't mind either way)

**Comment by kaxil** [comment]
> Somewhere I'd like to see us do a "static" deserialization that includes a Links - both ones defined in the operator, and one registered via a plugin, like we have with `serialized_simple_dag_ground_truth`.
> 
> (it might be time to move those to JSON files rather than having them in the test file? Don't mind either way)

Added

File: airflow/serialization/serialized_baseoperator.py (hunk 5 above)
Comment (ashb) [reply airflow/serialization/serialized_baseoperator.py]: What if someone wants to add a link via plugin that needs some attributes stored when (de)serializing?

File: airflow/contrib/operators/qubole_operator.py (hunk 1 above)
Comment (ashb) [reply airflow/contrib/operators/qubole_operator.py]: Oh hang on - I've got an idea. So if you delcare a var called `__x` python automatically converts that to `_MyClass__x`, so we can do this:

```python
class Base:
    def __init__(self, x=1):
        self.x = x

    __serialized_fields = None

    @classmethod
    def get_serialized_fields(cls):
        if not cls.__serialized_fields:
            cls.__serialized_fields = frozenset(vars(Base()))
        return cls.__serialized_fields


class  Sub(Base):
    def __init__(self, y, **kwargs):
        super().__init__(**kwargs)
        self.y = y

    __serialized_fields = None
    @classmethod
    def get_serialized_fields(cls):
        if not cls.__serialized_fields:
            cls.__serialized_fields = super().get_serialized_fields() | frozenset({'y'})
        return cls.__serialized_fields


print(Sub.get_serialized_fields())
print(Base.get_serialized_fields())
```

And then get as output

```
frozenset({'x', 'y'})
frozenset({'x'})
```

https://docs.python.org/3/tutorial/classes.html?highlight=class%20private#tut-private

File: airflow/serialization/serialized_objects.py [hunk 6]
```diff
@@ -202,7 +203,11 @@ def _deserialize(cls, encoded_var):  # pylint: disable=too-many-return-statement
             return [cls._deserialize(v) for v in encoded_var]
 
         assert isinstance(encoded_var, dict)
-        var = encoded_var[Encoding.VAR]
+        try:
+            var = encoded_var[Encoding.VAR]
+        except KeyError as e:
+            logging.info("encoded_var: %s", encoded_var)
```
Comment (ashb) [inline airflow/serialization/serialized_objects.py]: What's this line for? (Being at info seems like the wrong level)

File: tests/serialization/test_dag_serialization.py [hunk 7]
```diff
@@ -381,6 +362,49 @@ def test_roundtrip_relativedelta(self, val, expected):
...
+        simple_task = dag.task_dict["simple_task"]
+        self.assertEqual(getattr(simple_task, "bash_command"), "true")
+
+        #########################################################
+        # Verify Operator Links work with Serialized Operator
+        #########################################################
+        # Check Serialized version of operator link only contains the inbuilt Op Link
+        self.assertEqual(
+            serialized_dag["dag"]["tasks"][0]["_operator_links_sources"],
+            [{'airflow.utils.tests.CustomBaseOpLink': {}}]
```
Comment (ashb) [inline tests/serialization/test_dag_serialization.py]: What (if anything) is stored in the value (`{}`) -- is that are attributes of the operator link would go?

Comment (kaxil) [reply tests/serialization/test_dag_serialization.py]: Yes like:

https://github.com/apache/airflow/pull/6715/files#diff-5a167b261e34f0a91d371b5429ace188R492-R499 :

```
        # Check Serialized version of operator link
        self.assertEqual(
            serialized_dag["dag"]["tasks"][0]["_operator_links_sources"],
            [
                {'airflow.gcp.operators.bigquery.BigQueryConsoleIndexableLink': {'index': 0}},
                {'airflow.gcp.operators.bigquery.BigQueryConsoleIndexableLink': {'index': 1}}
            ]
        )
```

File: airflow/models/baseoperator.py [hunk 8]
```diff
@@ -1081,21 +1082,22 @@ def get_extra_links(self, dttm, link_name):
...
 
 
+@attr.s(auto_attribs=True)
 class BaseOperatorLink(metaclass=ABCMeta):
     """
     Abstract base class that defines how we get an operator link.
     """
 
-    operators: List[Type[BaseOperator]] = []
+    operators = []   # type: List[Type[BaseOperator]]
```
Comment (ashb) [inline airflow/models/baseoperator.py]: Don't think you wanted to "revert" this change when resolving conflicts.

File: tests/gcp/operators/test_bigquery.py [hunk 9]
```diff
@@ -432,6 +433,93 @@ def test_bigquery_operator_defaults(self, mock_hook):
...
+        with self.dag:
+            BigQueryOperator(
+                task_id=TASK_ID,
+                sql=['SELECT * FROM test_table', 'SELECT * FROM test_table2'],
+            )
+        serialized_dag = SerializedDAG.to_dict(self.dag)
+        self.assertIn("sql", serialized_dag["dag"]["tasks"][0])
+
+        dag = SerializedDAG.from_dict(serialized_dag)
+        simple_task = dag.task_dict[TASK_ID]
```
Comment (ashb) [inline tests/gcp/operators/test_bigquery.py:483]:
```suggestion
        deserialized_task = dag.task_dict[TASK_ID]
```

As we've currently got something else called SimpleTask and this isn't one of those.

File: airflow/serialization/serialized_objects.py [hunk 10]
```diff
@@ -350,6 +368,53 @@ def _is_excluded(cls, var, attrname, op):
...
+        """
+        Serialize Operator Links. Store the import path of the OperatorLink and the arguments
+        passed to it. Example ``[{'airflow.gcp.operators.bigquery.BigQueryConsoleLink': {}}]``
+
+        :param operator_extra_links: Operator Link
+        :return: Serialized Operator Link
+        """
+        return [
+            {
+                "{}.{}".format(x.__class__.__module__, x.__class__.__name__): cattr.unstructure(x)
```
Comment (ashb) [inline airflow/serialization/serialized_objects.py]: I wonder if we need to pass the RHS of this through `cls.serialize_to_json` -- otherwise more complex types than int/str/bool/dict/array can't be serialized? Or should we just expect that cattr/attr has handled that for us?

File: airflow/serialization/serialized_objects.py (hunk 6 above)
Comment (kaxil) [reply airflow/serialization/serialized_objects.py]: Removed

File: airflow/models/baseoperator.py (hunk 8 above)
Comment (kaxil) [reply airflow/models/baseoperator.py]: Fixed

File: tests/gcp/operators/test_bigquery.py (hunk 9 above)
Comment (kaxil) [reply tests/gcp/operators/test_bigquery.py:483]: It is just a task_id in tests specifically for Serialised Tests :D , does it matter? Happy to change the name but there are several places where the task_id is named simple_task. 

And I think we have SimpleTaskInstance and SimpleDagBag (that are soon to be removed too)

File: airflow/serialization/serialized_objects.py (hunk 10 above)
Comment (kaxil) [reply airflow/serialization/serialized_objects.py]: Updated the logic :)

File: airflow/contrib/operators/qubole_operator.py (hunk 1 above)
Comment (kaxil) [reply airflow/contrib/operators/qubole_operator.py]: Applied

Comment (kaxil) [reply airflow/contrib/operators/qubole_operator.py]: Didn't know that this kind of thing existed :D 

Thanks @ashb

File: airflow/serialization/serialized_objects.py [hunk 11]
```diff
@@ -354,6 +371,88 @@ def _is_excluded(cls, var: Any, attrname: str, op: BaseOperator):
...
+
+        for _operator_links_source in encoded_op_links:
+            _operator_link_class, data = list(_operator_links_source.items())[0]
+
+            single_op_link_class_name = None
+
+            try:
+                single_op_link_class_name = import_string(_operator_link_class)
+            except ImportError:
+                logging.debug("Module not found or not accessible: %s", _operator_link_class)
```
Comment (ashb) [inline airflow/serialization/serialized_objects.py]: My gut tells me we don't actually need any "else". Cos if the plugin is imported (which it will be) then and the class name comes from `__module__` and `__class__` then the module will already be imported.

File: airflow/utils/tests.py [hunk 12]
```diff
@@ -71,15 +75,70 @@ class Dummy3TestOperator(BaseOperator):
     operator_extra_links = ()
 
 
+@attr.s(auto_attribs=True)
+class CustomBaseIndexOpLink(BaseOperatorLink):
```
Comment (ashb) [inline airflow/utils/tests.py:79]:
```suggestion
class CustomIndexOpLink(BaseOperatorLink):
```

please -- it's not a Base class :)

File: airflow/utils/tests.py [hunk 13]
```diff
@@ -71,15 +75,70 @@ class Dummy3TestOperator(BaseOperator):
...
+        search_queries = ti.xcom_pull(task_ids=operator.task_id, key='search_query')
+        if not search_queries:
+            return None
+        if len(search_queries) < self.index:
+            return None
+        search_query = search_queries[self.index]
+        return 'https://console.cloud.google.com/bigquery?j={}'.format(search_query)
+
+
+class CustomBaseOpLink(BaseOperatorLink):
```
Comment (ashb) [inline airflow/utils/tests.py]:
```suggestion
class CustomOpLink(BaseOperatorLink):
```

File: tests/serialization/test_dag_serialization.py [hunk 14]
```diff
@@ -381,6 +362,177 @@ def test_roundtrip_relativedelta(self, val, expected):
...
+                "tasks": [
+                    {
+                        "task_id": "simple_task",
+                        "_downstream_task_ids": [],
+                        "_inlets": [],
+                        "_outlets": [],
+                        "_operator_extra_links": [
+                            # Intentionally wrong module name is added
+                            # to mimic the case where the webserver wont have access
+                            # to the OperatorLink code
```
Comment (ashb) [inline tests/serialization/test_dag_serialization.py]: To expand on my other comment: This can only happen when the webserver doesn't have the plugin either -- cos if a Link is defined in a plugin, then the normal plugin loading mechanism will (I think?) load the module under the same python module name (same key in `sys.modules`) so the import by class name will work on both.

That said: we need to be _very_ careful about this if it ever gets accepted from user/API input -- this sort of thing is a possible recipie for instantiating arbitrary python classes, which could lead to Vulns.

Comment (kaxil) [reply tests/serialization/test_dag_serialization.py]: Agreed, updated with the help from Ash himself :)

File: airflow/serialization/serialized_objects.py [hunk 15]
```diff
@@ -354,6 +370,65 @@ def _is_excluded(cls, var: Any, attrname: str, op: BaseOperator):
...
+
+        :param encoded_op_links: Serialized Operator Link
+        :return: De-Serialized Operator Link
+        """
+        from airflow.plugins_manager import registered_operator_link_classes
+
+        op_predefined_extra_links = {}
+
+        for _operator_links_source in encoded_op_links:
+            _operator_link_class, data = list(_operator_links_source.items())[0]
```
Comment (ashb) [inline airflow/serialization/serialized_objects.py:404]: We should probably add this to our JSON Schema. Something like this is the part of the schema we want for op_links:

```json
{ "type": "array",
  "items": {
    "type": "object",
    "minProperties": 1,
    "maxProperties": 1
  }
}
```

Comment (ashb) [reply airflow/serialization/serialized_objects.py:404]: This line took me a bit to work out what it's doing -- It modifies it in place, but `_operator_link_class, data = _operator_links_source.popitem()` works and is a bit clearer to me. What do you think?
(Don't know if that works for py2, which isn't a problem here but is when backporting this)

File: airflow/serialization/serialized_objects.py [hunk 16]
```diff
@@ -354,6 +370,65 @@ def _is_excluded(cls, var: Any, attrname: str, op: BaseOperator):
...
+
+            if _operator_link_class in registered_operator_link_classes:
+                single_op_link_class_name = registered_operator_link_classes[_operator_link_class]
+            else:
+                raise KeyError("Operator Link Not Found ..., ", _operator_link_class)
```
Comment (ashb) [inline airflow/serialization/serialized_objects.py]:
```suggestion
                raise KeyError("Operator Link class %r not registered" % _operator_link_class)
```

File: airflow/serialization/serialized_objects.py (hunk 15 above)
Comment (kaxil) [reply airflow/serialization/serialized_objects.py:404]: Done

**Review by ashb — APPROVED** [review]
